### ISSUE TITLE
Stop logging based on disk space

### ISSUE THREAD

**Issue by DanAvni** [serilog/serilog-sinks-file#200]
Please add the ability to stop logging to disk if free disk space falls below some threshold. I recently had a situation where disk space ran to 0 because logging kept on writing (file size limit is not set for various reasons and even if it did in this case it would still drive the disk to 0 as the limit would be higher than the available disk space). I know this is an extremely rare situation but still logging should not drive the free space to 0

**Comment by cocowalla** [comment]
Hi, appreciate you sharing a production issue from the trenches 😃 

I don't know what @nblumhardt thinks, but IMO I don't think it's up to the file sink to check for disk space. To do so would add complexity and possibly hit performance too. The "proper" solution here is likely to set up monitoring, so you (or an Ops team) receive alerts when disk space runs low. I realise that's a PITA if you're not already doing it, but it's worthwhile for a number of reasons, and is a standard solution to problems like these.

**Comment by nblumhardt** [comment]
Hi folks! I don't think we should tackle disk space checking in this sink, though it's a bit intriguing. The recommended option right now is to use `fileSizeLimitBytes` and `fileCountLimit` to achieve similar results.

I've written code along those lines in the past, and recall that getting a measurement of free disk space can be an expensive operation on some OSes.

Still, this would be a quick fix if you are keen:

```
    .Filter.ByExcluding(_ => GetFreeDiskSpace("...") < limit)
```

**Comment by cocowalla** [comment]
> Still, this would be a quick fix if you are keen:
> 
> ```
>     .Filter.ByExcluding(_ => GetFreeDiskSpace("...") < limit)
> ```

@DanAvni, if you really did want to go down this route, you could improve upon @nblumhardt's rather clever solution by not calculating free disk space on every emit (which I'm fairly sure could kill perf) - `GetFreeDiskSpace` could be an accessor for a variable that's updated periodically by a background thread/task.

And as another alternative to full-scale server monitoring, you could use something like [HealthChecks](https://github.com/Xabaril/AspNetCore.Diagnostics.HealthChecks).

**Comment by DanAvni** [comment]
@cocowalla, While I agree there is a performance hit, I do think it is required. We do have monitoring and it did alert when disk space was getting low but by the time I got to look, it was already 0. The issue was that someone was attacking our site causing many errors to be logged so the disk ran out of space. This happened on a test machine (so by our design it had little free disk space to begin with) and no real damage was done but should this happen in prod machines it could be problematic. 

I could easily imagine a DoS attack on a site by sending tons of invalid requests hoping that disk/memory fill up (see for example "Application-layer attacks" on https://en.wikipedia.org/wiki/Denial-of-service_attack ). And since this could happen fast I do think that logging should not be the reason a site is brought down.

Also, I know that in theory using the `fileSizeLimitBytes` should solve this but in (my) reality server free space changes and a limit I place today might not be valid tomorrow so a dynamic limit is much better in my eyes.

I do think the best implementation that will not affect performance as much would be by adding three properties to the Sink:
Drive letter to check for free space
The interval at which to sample free disk space (Could be 0 to always sample and every other timespan)
Free space threshold
That way free disk space can be measured every 10 seconds and assuming I set the threshold at 10GB that means I also assume no logging would use disk by 10GB of logs in 10 seconds. That will not prevent the running out of disk space completely but will lower the chances of a DoS attack actually bringing the server down.

@nblumhardt , As I am using Serilog from an XML config and in it I am writing to both a file and to Seq, having a filter would mean I would stop having events in Seq as well which is not the intended behavior. I saw I could create two child loggers and have only one (file sink) with a filter but also saw that this is not doable at the moment in Serilog XML config (see https://github.com/tsimbalar/serilog-settings-comparison/blob/master/docs/README.md#sub-loggers--child-loggers ). Being able to combine the sub loggers with the expressions package I saw (and adding to it some ability to sample disk space) could also be sort of solution although I am not sure it is the solution I would prefer.

Anyway, just my thoughts. I think I will also explore some way of de-deplicating exceptions log entries if they happen too frequently (so the log will log the 1st exception and if the same exception happens multiple times again it would with exponential backoff write event logs similar to "The following exception happened 100 times in the past 5 seconds <exception>", Then for 10 seconds, then 20 and so on. Until the exception stops and things reset to normal. Not sure this is applicable to all exceptions but for a specific sub-set of things that could happen in a loop it might be
